web/vtadmin/src/api/http.ts

```diff
@@ -15,68 +15,65 @@
  */
 
 import { vtadmin as pb } from '../proto/vtadmin';
+import * as errorHandler from '../errors/errorHandler';
+import {
+    HttpFetchError,
+    HttpResponseNotOkError,
+    MalformedHttpResponseError,
+ } from '../errors/errorTypes'
+import { HttpOkResponse } from './responseTypes';
 
-interface HttpOkResponse {
-    ok: true;
-    result: any;
-}
-
-interface HttpErrorResponse {
-    ok: false;
-}
-
-export const MALFORMED_HTTP_RESPONSE_ERROR = 'MalformedHttpResponseError';
-
-// MalformedHttpResponseError is thrown when the JSON response envelope
-// is an unexpected shape.
-class MalformedHttpResponseError extends Error {
-    responseJson: object;
-
-    constructor(message: string, responseJson: object) {
-        super(message);
-        this.name = MALFORMED_HTTP_RESPONSE_ERROR;
-        this.responseJson = responseJson;
-    }
-}
-
-export const HTTP_RESPONSE_NOT_OK_ERROR = 'HttpResponseNotOkError';
-
-// HttpResponseNotOkError is throw when the `ok` is false in
-// the JSON response envelope.
-class HttpResponseNotOkError extends Error {
-    response: HttpErrorResponse | null;
-
-    constructor(endpoint: string, response: HttpErrorResponse) {
-        super(endpoint);
-        this.name = HTTP_RESPONSE_NOT_OK_ERROR;
-        this.response = response;
-    }
-}
-
-// vtfetch makes HTTP requests against the given vtadmin-api endpoint
-// and returns the parsed response.
-//
-// HttpResponse envelope types are not defined in vtadmin.proto (nor should they be)
-// thus we have to validate the shape of the API response with more care.
-//
-// Note that this only validates the HttpResponse envelope; it does not
-// do any type checking or validation on the result.
+/**
+ * vtfetch makes HTTP requests against the given vtadmin-api endpoint
+ * and returns the parsed response.
+ *
+ * HttpResponse envelope types are not defined in vtadmin.proto (nor should they be)
+ * thus we have to validate the shape of the API response with more care.
+ *
+ * Note that this only validates the HttpResponse envelope; it does not
+ * do any type checking or validation on the result. 
+ */
 export const vtfetch = async (endpoint: string): Promise<HttpOkResponse> => {
-    const { REACT_APP_VTADMIN_API_ADDRESS } = process.env;
-
-    const url = `${REACT_APP_VTADMIN_API_ADDRESS}${endpoint}`;
-    const opts = vtfetchOpts();
-
-    const response = await global.fetch(url, opts);
-
-    const json = await response.json();
-    if (!('ok' in json)) throw new MalformedHttpResponseError('invalid http envelope', json);
-
-    // Throw "not ok" responses so that react-query correctly interprets them as errors.
-    // See https://react-query.tanstack.com/guides/query-functions#handling-and-throwing-errors
-    if (!json.ok) throw new HttpResponseNotOkError(endpoint, json);
-
-    return json as HttpOkResponse;
+    try {
```

Comment:
Unfortunately this diff looks much worse than it is, since I had to surround everything in a `try/catch`. Logic changes here are minimal and well-tested. 🙏